Add initial opsgenie plugin RFD (#22221)
gravitational/teleport: rfd/0109-opsgenie-plugin.md

# Opsgenie Integration RFD
## Required approvers

Engineering:
Product:
Security:

## What

This RFD proposes a plugin that allows Teleport to integrate with Opsgenie, allowing access requests to show up as alerts in Opsgenie. This plugin will differ from existing plugins by being part of the Teleport binary directly.

## Success criteria
Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
Alerts were chosen over incidents in Opsgenie for high priority alerts that indicate a service interruption.
Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call 

## Configuration UX

> The plugin will be configured using a toml file containing the required information to interact with both Teleport access and the Opsgenie API.

Reviewer: Let's try to avoid adding [two competing plugin configuration mechanisms](https://github.com/gravitational/teleport/pull/20990) in Teleport at the same time if possible.

Reviewer 2: Agreed :+1:

Reviewer 3: Hard agree. However, this needs some thought, as the scope of this configuration in hosted and standalone versions will be different:
* Hosted: we only configure what Cloud needs to authenticate with the provider (in OAuth case - client credentials; in OpsGenie case - might be nothing, since the user will need to supply a static API token). All the "settings knobs" are controled via:
  * `Plugin` resource which can be manipulated through UI (and UI-only for now at least),
  * Eventually, routing rules.
*  Self-hosted standalone: we need to configure Teleport identity & address, 3rd party provider auth, as well as "settings" (OpsGenie: priority, tags, etc.).

```
[opsgenie]
api_key = "path/to/key" # File containing Opsgenie API Key
opsgenie_addr = "example.app.opsgenie.com" # Address of Opsgenie
severity = "2" # Severity to create Opsgenie alerts with

> ```

Author: I think at least for now we should leave out the ability to run multiple instances in this way. Since it can be run as a standalone application users can always just run it twice that way and pass each instance a different config file.

### Getting an Opsgenie API key

In the Opsgenie web UI go to Settings -> App settings -> API key management. Create key with Read, Create and Update access.

### Executing

The plugin will be started using a command of the form

```
Teleport opsgenie start –config <config file location>
```

## UX

Once an access request has been created, the Opsgenie plugin will create an alert in the service specified in the request annotation using the Opsgenie Alert API Create endpoint. 

> The appropriate on call responder can then click into the provided link and approve or deny the access request.

Reviewer: Will this link to the web UI for the access requests? In the past we've had security issues with links that directly approved or denied access requests in plugins.

Reviewer 2: Somewhat related.

The JIRA plugin updates the access request based on jira events (webhooks) when the user moves the ticket (using the JIRA Web App) to Approve or Deny column.

I'm not sure if this is possible with Opsgenie, but we should be clear about this: Opsgenie only reads Teleport state, and never writes to it.

There's only one thing that could trigger an approval: auto approval (which, IMO, should be configurable).
However, that isn't triggered by Opsgenie itself but from the plugin process

Author: That was the intention. Do you have any more info on the past security issues doing this and how they were avoided?

Reviewer 4: I think what @espadolini is referring to, is the link should take the user to the page for this access request which has approve/deny buttons the user should click explicitly. Rather than having a link that just auto-approves when you click on it (which is prone to phishing attacks). I don't even think we have ability to generate such links but might be worth calling this out explicitly in the RFD.

For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the services provided in request annotation.

Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.

## Implementation details
In this section we will take a look at how the plugin will interact with the Opsgenie API.

### Authorization

The plugin will use the API key provided in the Opsgenie config file when interacting with the Opsgenie API. This API key will be included in the headers of the requests made.

```
Header Key: Authorization
Header Value: genieKey $apiKey
```

### Creating alerts
When the Opsgenie plugin creates alerts for incoming access requests the Create alert request will be of the form

```
{
	"message": "Access request from <someuser>",
	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
 	Reason: <someReason>
 	To approve or deny the request, proceed to <link to the access request>",
	"responders":[
    	....
	],
	"tags": ["TeleportAccessRequest"],

> 	"priority":"<somePriority>"

Reviewer 2: Is it the severity that we have in the config?
If so, I think we should be consistent with the naming.

}
```

> When the access request has been approved or denied the alert created in Opsgenie will have a note added to it using the ‘Add note to alert’ endpoint. Then the alert will be closed using the ‘Close alert’ endpoint.

Reviewer: Mind you, it's possible to configure the access request system so that multiple reviews are required to approve or deny an access request; this should read something along the lines of "on every review, the alert will have a note added to it, and when the access request is approved or denied, the alert will be closed" (it's also possible to deny an access request that was already approved, or to delete an access request at any point).

```
<Reviewer> reviewed the request at <someTime>.
Resolution: <StateEmojiAsUsedByExistingPLugins> <State>.
Reason: <Reason>
```

### Auto approval

To check if the requesting user of a request is currently on-call the ‘Who is on call API’s ‘Get on calls’ endpoint will be used. 'https://<configured-opsgenie-address>/v2/schedules/<SheduleName>/on-calls?scheduleIdentifierType=name'

Similar to the existing Pagerduty plugin for auto-approval to work, the user creating an Access Request must have a Teleport username that is also the email address associated with an Opsgenie account.

Access requests will be mapped to Opsgenie alerts by including the Access request ID in the tags field of the note. 

Shared code between the teleport-plugins found in lib is not too extensive and the simplest method to handle this when adding the Opsgenie plugin would be to simply duplicate what is needed for now.

## Security considerations

Potential for users to get access requests auto approved if they can get themselves onto the current on call rotation.

> Since Teleport usernames are assumed to match the Opsgenie email address when checking on call there is potential for access requests to be auto approved unintentionally.

Reviewer 2: We could decrease the severity of this if we allow for the configuration of auto approval

> # OPSGenie Integration RFD

Reviewer 5: ```suggestion
# Opsgenie Integration RFD
```

Isn't this the correct capitalization?

## Required approvers

Engineering:
Product:
Security:

## What

> This RFD proposes a plugin that allows Teleport to integrate with OpsGenie. This plugin will differ from existing plugins by being part of the Teleport binary directly.

Reviewer 5: _integrate with Opsgenie_ is quite vague. Integrate to do what?

## Success criteria
Users are able to configure teleport to automatically create alerts in OpsGenie from access requests.

> Alerts were chosen over incidents in OpsGenie for high priority alerts that indicate a service interruption.

Reviewer 5: Could you elaborate more on this?

Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call 

## Configuration UX

> The plugin will be configured using a toml file containing the required to interact with both Teleport access and the OpsGenie API.

Reviewer 5: If the plugin is part of the Teleport binary, why did we decide to use a separate TOML file as opposed to the existing teleport.yaml configuration file?

```
[teleport]

> addr = "example.com:3025" # Teleport Auth Server GRPC API address

Reviewer 5: If the plugin is running in the same process as the auth server, why do we need to give it an addr? Won't this just make it harder to configure correctly?

Author: Removed these.

Reviewer 4: @zmb3 @EdwardDowling I think we want to maintain the ability to run the plugin separately, not just as part of the auth server. Similar to how our other access plugins function. For example `teleport plugin opsgenie start`, `teleport plugin pagerduty start`, etc.

Reviewer 2: If we end up using the `teleport.yaml` format, then I guess we should support the current `teleport start --config teleport.yaml` flow.

client_key = "/var/lib/teleport/plugins/opsgenie/auth.key" # Teleport GRPC client secret key
client_crt = "/var/lib/teleport/plugins/opsgenie/auth.crt" # Teleport GRPC client certificate
root_cas = "/var/lib/teleport/plugins/opsgenie/auth.cas" # Teleport cluster CA certs

[opsgenie]

> api_key = "key" # Opsgenie API Key

Reviewer 5: Did you consider having this be a path to a file so that we don't have to put secrets in the configuration file?

[log]

> output = "stderr" # Logger output. Could be "stdout", "stderr" or "/var/lib/teleport/opsgenie.log"

Reviewer 5: Similar to above comments - if the plugin is running in-process, I would expect it to use the same logging config as the rest of the process.

severity = "INFO" # Logger severity. Could be "INFO", "ERROR", "DEBUG" or "WARN".
```

### Getting an OpsGenie API key

In the OpsGenie web UI go to Settings -> App settings -> API key management. Create key with Read, Create and Update access.

### Executing

The plugin will be started using a command of the form

```
Teleport opsgenie start –config <config file location>
```

## UX

> Once an access request has been created, the OpsGenie plugin will create an alert in service specified in the config file using the OpsGenie Alert API Create endpoint. 

Reviewer 5: Does this mean that submitting an access request is going to page someone? Is that what we want?

Author: I've added an option to specify the severity to the config file so users can use that to determine whether or not the alerts will be created at a level that pages someone.

For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the services provided in request annotation.

Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.

## Implementation details
In this section we will take a look at how the plugin will interact with the OpsGenie API.

### Authorization

The plugin will use the API key provided in the OpsGenie config file when interacting with the OpsGenie API. This API key will be included in the headers of the requests made.

```
Header Key: Authorization
Header Value: GenieKey $apiKey
```

### Creating alerts
When the OpsGenie plugin creates alerts for incoming access requests the Create alert request will be of the form

```
{
	"message": "Access request from <someuser>",
	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
 	Reason: <someReason>
 	To approve or deny the request, proceed to <link to the access request>",
	"responders":[
    	....
	],

> 	"tags": ["TeleportAccessRequest"],

Reviewer 5: What purpose do the tags serve? Should we allow users to configure what tags are placed on the alert?

Author: Added this just so its easier to filter the alerts created by Teleport in the Opsgenie ui. I'm open to the idea of adding in an option to configure the tags added. if you think users might want that.

Reviewer: Users with multiple clusters might want to have that in the tags as well, we should leave the tags up to the user.

# Opsgenie Integration RFD
## Required approvers

> Engineering:
> Product:
> Security:

Reviewer 4: ```suggestion
Engineering: @r0mant, @marcoandredinis, @hugoShaka
Product: @klizhentas, @xinding33
Security: @reedloden, @jentfoo 
```

## What

This RFD proposes a plugin that allows Teleport to integrate with Opsgenie, allowing access requests to show up as alerts in Opsgenie. This plugin will differ from existing plugins by being part of the Teleport binary directly.

## Success criteria
Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
Alerts were chosen over incidents in Opsgenie as incidents are intended to be used for high priority alerts that indicate a service interruption.

> Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call 

Reviewer 4: ```suggestion
Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call.
```

## Configuration UX

The plugin will be configured in Teleport's config yaml file. The required fields will be added to a 'plugins' section containing the required information to interact with both Teleport access and the Opsgenie API.

```
plugins:

>     opsgenie:

Reviewer 4: When the plugin runs standalone, how does it connect to the cluster? It should have auth address and identity file in the settings, right? If we're using teleport.yaml fileconfig format, there's already auth_server field but we should be able to pass it identity somehow.

Reviewer 4: Also, if we're running inside Teleport, can the plugin be just a separate service (like, app service, database service, etc.) and retrieve its identity same way as any other agent? cc @espadolini what do you think?

Reviewer 2: I like the idea of having this to be a separate service :+1: 
We would need to create a join token and a new System Role for plugins
That would decrease the setup time because you no longer need to create a user and a role to interact with the API

Reviewer 2: :nail_care:
Should we allow for multiple instances of the same plugin?
I don't think this would be required now but could allow us to have a generic HTTP plugin in the future that works with multiple endpoints.

        api_key: "path/to/key" # File containing Opsgenie API Key

>         opsgenie_addr: "example.app.opsgenie.com" # Address of Opsgenie

Reviewer 4: "opsgenie" prefix seems redundant.
```suggestion
        addr: "example.app.opsgenie.com" # Address of Opsgenie API endpoint
```

        priority: "2" # Priority to create Opsgenie alerts with
        alert_tags: ["example-tag"] # List of tags to be added to alerts created in Opsgenie

>         auto_approval: true # Whether or not to enable auto approval

Reviewer 4: For PagerDuty plugin auto-approval is done by having an annotation on the user's role that specifies services the user is on-call in. Why do we need this flag for OpsGenie?

Reviewer 2: Sorry, this was probably added because I commented that auto-approval should be configurable.
PagerDuty AutoApproval is not configurable, so feel free to ignore my previous comment.

Author: I'll remove this flag then and handle this the same way pagerduty plugin does

```

The logging configuration will be shared with the main Teleport process.

### Getting an Opsgenie API key

In the Opsgenie web UI go to Settings -> App settings -> API key management. Create key with Read, Create and Update access.

### Executing

The plugin will be started using a command of the form

```

> teleport start --config /etc/teleport.yaml

Reviewer 4: So the plugin will be just another Teleport service, correct? And you can start it alongside other agents like SSH, or auth or proxy or anything else, just like any other agent?

```

## UX

> Once an access request has been created, the Opsgenie plugin will create an alert in the service specified in the request annotation using the Opsgenie Alert API Create endpoint. 

Reviewer 4: Can you give a role example with this annotation.

The appropriate on call responder can then click into the provided link and approve or deny the access request.

> For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the services provided in request annotation. (Provided auto approval is enabled in the config)

Reviewer 4: Again, can you please include the role example with this annotation.

Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.

> ## Implementation details

Reviewer 4: In the implementation details, can you include a section that will explain how the plugin will be integrated in the teleport binary? Are we going to import stuff from teleport-plugins? I think @justinas mentioned we already do this for the hosted plugins work?

In this section we will take a look at how the plugin will interact with the Opsgenie API.

### Authorization

The plugin will use the API key provided in the Opsgenie config file when interacting with the Opsgenie API. This API key will be included in the headers of the requests made.

```
Header Key: Authorization

> Header Value: genieKey $apiKey

Reviewer 4: What are "Header Key" and "Header Value"? Let's be explicit.

```

### Creating alerts
When the Opsgenie plugin creates alerts for incoming access requests the Create alert request will be of the form

```
{
	"message": "Access request from <someuser>",
	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
 	Reason: <someReason>
 	To approve or deny the request, proceed to <link to the access request>",
	"responders":[
    	....
	],
	"tags": ["TeleportAccessRequest"],
	"priority":"<somePriority>"
}
```

> On every review of the access request the alert created in Opsgenie will have a note added to it using the ‘Add note to alert’ endpoint. Then the alert will be closed using the ‘Close alert’ endpoint once the alert has either been approved or denied.

Reviewer 4: Nit: Whenever you mention an OpsGenie API you intend to use, including a link to the API reference in their docs (or  just the API call itself) would be useful.

```
<Reviewer> reviewed the request at <someTime>.
Resolution: <StateEmojiAsUsedByExistingPLugins> <State>.
Reason: <Reason>
```

### Auto approval

> To check if the requesting user of a request is currently on-call the ‘Who is on call API’s ‘Get on calls’ endpoint will be used. 'https://<configured-opsgenie-address>/v2/schedules/<SheduleName>/on-calls?scheduleIdentifierType=name'

Reviewer 2: Should we add the `ScheduleName` to the config?

Similar to the existing Pagerduty plugin for auto-approval to work, the user creating an Access Request must have a Teleport username that is also the email address associated with an Opsgenie account.

Access requests will be mapped to Opsgenie alerts by including the Access request ID in the tags field of the note. 

Shared code between the teleport-plugins found in lib is not too extensive and the simplest method to handle this when adding the Opsgenie plugin would be to simply duplicate what is needed for now.

### Note

> Opsgenie will only ever read Teleport state and never write it.

Reviewer 4: This feels out of place without additional context. Do we need this note?

## Security considerations

Potential for users to get access requests auto approved if they can get themselves onto the current on call rotation.

> Since Teleport usernames are assumed to match the Opsgenie email address when checking on call there is potential for access requests to be auto approved unintentionally.

Reviewer 2: But only for the roles they are allowed to request access to, right?
As an example
https://goteleport.com/docs/access-controls/access-requests/role-requests/#role-access-requests-setup
In this case, the user is only allowed to ask for the `dba` role

Author: Yes, this would only be a concern for roles they are able to create access requests for.

## Scope

### Phase 1

The plugin will be run as a standalone application similar to the existing Pagerduty plugin differing only in that it will be part of the Teleport binary.

### Phase 2

The plugin will follow the 'hosted' operations model and will at that point be possible to run as part of the Auth server directly or as a standalone application.

## Success criteria
Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
Alerts were chosen over incidents in Opsgenie as incidents are intended to be used for high priority alerts that indicate a service interruption.
Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call.

## Configuration UX

The plugin will be configured in Teleport's config yaml file. The required fields will be added to a 'plugins' section containing the required information to interact with both Teleport access and the Opsgenie API.

```

> plugins:

Reviewer 2: Other plugins currently support more fields
- log level/severity
- log output location
- Auth Server
- Identity or ClientKey/Cert and RootCAs

Should we add them within the `opsgenie`, `plugins` or a top level `teleport` key? 

Given that Identity is already at `opsgenie`, I think we should add the ClientKey/Cert and RootCAs to it as well.
However, it could make sense to add them as some kind of generic plugins configuration (we can always do that later on :+1: )

Reviewer 2: > Was planning for the logging config to be shared with the rest of the teleport config

Something like this?
```yaml
teleport:
  log:
     level: INFO
     output: stderr
plugins:
   opsgenie:
      addr: ...

Author: That was what i was thinking.

```
Header Key: Authorization
Header Value: genieKey $apiKey
```

### Creating alerts
When the Opsgenie plugin creates alerts for incoming access requests the [Create](https://docs.opsgenie.com/docs/alert-api#create-alert) alert request will be of the form

```
{
	"message": "Access request from <someuser>",
	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
 	Reason: <someReason>
 	To approve or deny the request, proceed to <link to the access request>",
	"responders":[
    	....
	],
	"tags": ["TeleportAccessRequest"],
	"priority":"<somePriority>"
}
```

On every review of the access request the alert created in Opsgenie will have a note added to it using the ‘[Add note to alert](https://docs.opsgenie.com/docs/alert-api#add-note-to-alert)’ endpoint. Then the alert will be closed using the ‘[Close alert](https://docs.opsgenie.com/docs/alert-api#close-alert)’ endpoint once the alert has either been approved or denied.

```
<Reviewer> reviewed the request at <someTime>.
Resolution: <StateEmojiAsUsedByExistingPLugins> <State>.
Reason: <Reason>
```

### Auto approval

> To check if the requesting user of a request is currently on-call the ‘Who is on call API’s ‘[Get on calls](https://docs.opsgenie.com/docs/who-is-on-call-api#get-on-calls)’ endpoint will be used. 'https://<configured-opsgenie-address>/v2/schedules/<SheduleName>/on-calls?scheduleIdentifierType=name'

Reviewer 2: Where do we get the ScheduleName?
Should it be part of the opsgenie config?

Author: Had not considered this. We could add it to the config but maybe it would make more sense for this to be part of the request annotations since the service is already there.

Reviewer 2: I think a default value in config and then request annotations if override is necessary

Well, up to you where we put them :+1:

## Success criteria

### Plugin service
The creation of a Teleport service where plugins can be created as configurable resources similar to the existing 'app_service' and 'db_service'.

Plugins managed through this service should have the ability to be enabled and disabled dynamically via 'tctl'.

### Opsgenie plugin
Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
Alerts were chosen over incidents in Opsgenie as incidents are intended to be used for high priority alerts that indicate a service interruption.
Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call.

## Configuration UX

The plugin service will be configured using the existing Teleport YAML file in a section called 'plugin_service'.

```
plugin_service:
    enabled: true
    plugins:
    - "type": "opsgenie"
    - "type": "pagerduty"
    opsgenie:
        api_key: "path/to/key.txt" # Path to a file containing the Opsgenie API key
```
This example would only match plugin resources with the the labels 'type:opsgenie', or 'type:slack'.

> The Opsgenie plugin (and any others created) can then be configured using resources. 

Reviewer 4: Can you add a section about IaC as well? Basically, say that Terraform provider, Kubernetes operator will be updated to support managing the Plugin resource and Helm charts will be updated to support plugin_service.

And add it to the tracking ticket too please.

Example plugin.yaml for Opsgenie that would match with this.
```

> kind: plugin

Reviewer 4: @justinas I think you're introducing Plugin resource as part of hosted plugins work? Can you review this Plugin resource spec and see if it's consistent with how you designed it?

metadata:
  name: opsgenie-plugin
  labels:
    type: opsgenie
spec:

>   opsgenie:

Reviewer 3: Can one run several OpsGenie plugins using different `api_key`-s? If so, it should be possible to specify the auth secret (API key) in the Plugin resource (Plugin.[credentials](https://github.com/gravitational/teleport/blob/1c3188a38a312d7f85153d172ed52a1dfb0ad996/api/proto/teleport/legacy/types/types.proto#L4921-L4923)) rather than globally (?) in `plugins_service` config.

In any case, this would be a prerequisite for OpsGenie plugin in Cloud, as the users would need to supply the API key in some way, and they can not edit the config in Cloud.

Author: At the moment was not considering multiple API keys. We wanted to avoid putting secrets in the resource.

Reviewer 4: Yeah, we'd like to avoid secrets in the resource. This would likely means that we won't be able to support OpsGenie in the Cloud in the fully-hosted way as users will have to deploy an agent that has an API key but that's fine for now (it was discussed with the product team).

>     addr: "example.app.opsgenie.com" # Address of Opsgenie

Reviewer 3: Is there OpsGenie on-prem, or is this for use with mock servers in debug scenarios? Slack does not have such option currently, and for debugging we might instead want to have some environment var instead?

Author: There are multiple instances of Opsgenie that people might be using. The usual  api.opsgenie.com and the api.eu.opsgenie.com 
While this could be used for debug instances it wasn't the main intention.

    default_schedules: ["schedule1"] # Default on-call schedules to check if none are provided in the access request annotations
```

Given the above example teleport.yaml the following plugin resource will not match.
```
kind: plugin
metadata:
  name: slack-plugin
  labels:
    type: slack
spec:
  slack:
    addr: "example.slack.com"
```

The logging configuration will be shared with the main Teleport process.

### Getting an Opsgenie API key

In the Opsgenie web UI go to Settings -> App settings -> API key management. Create key with Read, Create and Update access.

### Executing

The plugin will be started using a command of the form given the appropriate 'plugin_service' is enabled in the config.

```
teleport start --config /etc/teleport.yaml
```

## UX

### Opsgenie plugin
Once an access request has been created, the Opsgenie plugin will create an alert in the service specified in the request annotation using the Opsgenie Alert API [Create](https://docs.opsgenie.com/docs/alert-api#create-alert) endpoint. 

The appropriate on-call responder can then click the provided link to the access request and approve or deny it.

> For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the services provided in access request annotations.

Reviewer 4: > services provided in access request annotations

Be explicit about which annotation this will be.

Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.

Example role with services in the annotation that indicate that users with this roll can be on-call for those services.

```
kind: role
metadata:
  name: someRole
spec:
  allow:
    request:
      roles: [someOtherRole]
      annotations:

>         opsgenie_services: ["service1", "service2"] # These are the Opsgenie services alerts will be created under

Reviewer 4: PagerDuty plugin uses 2 annotations: `pagerduty_notify_service` to indicate the service where alert will be created and `pagerduty_services` to indicate the on-call services for auto-approval flow.

You only have `opsgenie_services` defined here currently - which one is that? Can we make it consistent with how PD plugin works?

Also, I'm not very clear on terminology. Below you say that we need to know the "schedule" for the on-call user - so is it a schedule or a service?

```

## Implementation details

### Plugin service
The 'plugin_service' when enabled will watch for 'plugin' resources matching the labels specified in teleport.yaml.
When a plugin resource matching these labels is created the appropriate plugin will be started.

> In the case of the Opsgenie plugin, this will fail in the event the API key file field is not filled in teleport.yaml

Reviewer 4: ```suggestion
In the case of the Opsgenie plugin, the plugin startup will fail in the event the API key file field was not provided in the plugin_service configuration.
```

### Executing

The plugin will be started using a command of the form given the appropriate 'plugin_service' is enabled in the config.

```
teleport start --config /etc/teleport.yaml
```

## UX

### Opsgenie plugin
Once an access request has been created, the Opsgenie plugin will create an alert in the service specified in the request annotation opsgenie_notify_services using the Opsgenie Alert API [Create](https://docs.opsgenie.com/docs/alert-api#create-alert) endpoint.

The appropriate on-call responder can then click the provided link to the access request and approve or deny it.

For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the schedules provided in access request annotations.

Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.

Example role with services in the annotation that indicate that users with this roll can be on-call for those services.

```
kind: role
metadata:
  name: someRole
spec:
  allow:
    request:
      roles: [someOtherRole]
      annotations:

>         opsgenie_notify_services: ["service1", "service2"] # These are the Opsgenie services alerts will be created under

Reviewer 6: @EdwardDowling @r0mant keep in mind that we want to deprecate all this in favor of https://github.com/gravitational/teleport/blob/master/rfd/0087-access-request-notification-routing.md in Q2 cc @justinas

> # Opsgenie Integration RFD

Reviewer 4: Nit: Use the same title format as all other RFDs.
```suggestion
# RFD 109: Opsgenie Integration
```

Btw, is it Opsgenie or OpsGenie? Very minor and purely cosmetic, but can we please make sure we're using one or the other consistently throughout the document?

## Required approvers

Engineering: @r0mant, @marcoandredinis, @hugoShaka
Product: @klizhentas, @xinding33
Security: @reedloden, @jentfoo

## What

This RFD proposes a plugin that allows Teleport to integrate with Opsgenie, allowing access requests to show up as alerts in Opsgenie. This plugin will differ from existing plugins by being part of the Teleport binary directly. 
Doing so by creating a new Teleport service 'plugin_service' with the opsgenie plugin (and any future ones) being configurable resources. This will allow the plugins to be run either by themselves or alongside other Teleport services.

## Scope

### Phase 1

The plugin will be run as a standalone application similar to the existing Pagerduty plugin differing only in that it will be part of the Teleport binary.

### Phase 2

> The plugin will follow the 'hosted' operations model and will at that point be possible to run as part of the Auth server directly or as a standalone application.

Reviewer 4: I think these phases are no longer accurate. I would update the scope to the following:

- The plugin will be able to run standalone or alongside other services as part of the new `plugin_service`.
- The plugin will support auto-approval similar to PagerDuty plugin.

## Success criteria

### Plugin service
The creation of a Teleport service where plugins can be created as configurable resources similar to the existing 'app_service' and 'db_service'.

Plugins managed through this service should have the ability to be enabled and disabled dynamically via 'tctl'.

### Opsgenie plugin
Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
Alerts were chosen over incidents in Opsgenie as incidents are intended to be used for high priority alerts that indicate a service interruption.
Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call.

## Configuration UX

The plugin service will be configured using the existing Teleport YAML file in a section called 'plugin_service'.

```
plugin_service:
    enabled: true

>     resources:

Reviewer 4: Let's rename to "plugins" to make it more clear.
```suggestion
    plugins:
```

    - "type": "opsgenie"
    - "type": "pagerduty"
    opsgenie:
        api_key: "path/to/key.txt" # Path to a file containing the Opsgenie API key
```
This example would only match plugin resources with the the labels 'type:opsgenie', or 'type:slack'.

The Opsgenie plugin (and any others created) can then be configured using resources. 
Example plugin.yaml for Opsgenie that would match with this.
```
kind: plugin
metadata:

>   name: opsgenie-plugin
>     labels:
>         type: opsgenie

Reviewer 4: ```suggestion
  name: opsgenie-plugin
  labels:
    type: opsgenie
```

spec:
  opsgenie:
    addr: "example.app.opsgenie.com" # Address of Opsgenie
    priority: "2" # Priority to create Opsgenie alerts with
    alert_tags: ["example-tag"] # List of tags to be added to alerts created in Opsgenie

>     default_shedules: ["shedule1"] # Default on-call shedules to check if none are provided in the access request annotations

Reviewer 4: ```suggestion
    default_schedules: ["schedule1"] # Default on-call schedules to check if none are provided in the access request annotations
```

```

Given the above example teleport.yaml the following plugin resource will not match.
```
kind: plugin
metadata:

>   name: slack-plugin
>     labels:
>         type: slack

Reviewer 4: ```suggestion
  name: slack-plugin
  labels:
    type: slack
```

```
Header Key: Authorization
Header Value: genieKey $apiKey
```

### Creating alerts
When the Opsgenie plugin creates alerts for incoming access requests the [Create](https://docs.opsgenie.com/docs/alert-api#create-alert) alert request will be of the form

```
{
	"message": "Access request from <someuser>",
	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
 	Reason: <someReason>
 	To approve or deny the request, proceed to <link to the access request>",
	"responders":[
    	....
	],
	"tags": ["TeleportAccessRequest"],
	"priority":"<somePriority>"
}
```

On every review of the access request the alert created in Opsgenie will have a note added to it using the ‘[Add note to alert](https://docs.opsgenie.com/docs/alert-api#add-note-to-alert)’ endpoint. Then the alert will be closed using the ‘[Close alert](https://docs.opsgenie.com/docs/alert-api#close-alert)’ endpoint once the alert has either been approved or denied.

```
<Reviewer> reviewed the request at <someTime>.
Resolution: <StateEmojiAsUsedByExistingPLugins> <State>.
Reason: <Reason>
```

### Auto approval

> To check if the requesting user of a request is currently on-call the ‘Who is on call API’s ‘[Get on calls](https://docs.opsgenie.com/docs/who-is-on-call-api#get-on-calls)’ endpoint will be used. 'https://<configured-opsgenie-address>/v2/schedules/<SheduleName>/on-calls?scheduleIdentifierType=name'

Reviewer 4: Do we need to know schedules? IIRC in PagerDuty plugin we just see who currently is on-call for a particular service. Does OpsGenie provide a similar API?

> Similar to the existing Pagerduty plugin for auto-approval to work, the user creating an Access Request must have a Teleport username that is also the email address associated with an Opsgenie account.

Reviewer 4: IIRC PagerDuty plugin is using a separate role annotation to indicate services eligible for auto-approval. Will OpsGenie follow the similar approach?

Author: Yes, the services specified in the annotations will be used for that alongside default services that can be specified in the plugin resource.

Access requests will be mapped to Opsgenie alerts by including the Access request ID in the tags field of the note. 

> Shared code between the teleport-plugins found in lib is not too extensive and the simplest method to handle this when adding the Opsgenie plugin would be to simply duplicate what is needed for now.

Reviewer 4: This feels out of place and unclear, remove it from the RFD?

The plugin will follow the 'hosted' operations model and will at that point be possible to run as part of the Auth server directly or as a standalone application.

## Success criteria

### Plugin service
The creation of a Teleport service where plugins can be created as configurable resources similar to the existing 'app_service' and 'db_service'.

> Plugins managed through this service should have the ability to be enabled and disabled dynamically via 'tctl'.

Reviewer 2: Let me see if I got this straight.
We'll create plugin resource instance like any other resource (via API or `tctl create myplugins.yaml`)
This only stores the plugin info in the backend.

When we start a `teleport.yaml` with a `plugin_service`, based on label matching, it will start the actual processing.

So, the plugin configuration lives within the `myplugins.yaml` and is stored in the Plugin resource kind.
We then need to configure a `teleport.yaml` with something like
```
plugin_service:
    opsgenie:
        resource_matcher:
             - '*': '*'
```
Then, we need to `teleport start --config teleport.yaml` and it will start the actual opsgenie plugin.

Did I get that right?

Reviewer 4: I would just use the same notation as app/db services tbh so the service will monitor plugins of any kind matching the labels. Users can control which service matches which plugins via labels.

```yaml
plugin_service:
  resources:
  - "type": "slack"
    "team": "teamA"
  - "type": "opsgenie"
    "team": "teamB"
```

Reviewer 2: One last question
If we start two instances using the same `resources` set, will it run twice? Or is there some way to prevent that?

Reviewer 2: Btw, why do we have two places to configure plugin params? We can change the team here and other settings on the plugin resource itself.
Is there some kind of inheritance? If configured in `plugin_service.resources.XYZ` then the value of `plugin.spec.XYZ` is discarded and we use the former instead

Reviewer 6: @r0mant does it even have to be a static config matcher? How would this work with cloud? Can we auto match based on annotation or type? What's the value?

Reviewer 6: @r0mant I get it, this is to turn on the plugin for specific service, then `resources` does not really describe it well. How can we make it more clear?

Reviewer 4: @klizhentas For the cloud I was thinking that the plugin_service would run together with auth_service and would have a wildcard labels selector so it would pick up any plugin the user registers. If "resources" is confusing, how about "plugins"?

```
plugin_service:
  enabled: yes
  plugins:
  - "*": "*"
```
